recipes/wirelessStatus.lua

```diff
@@ -0,0 +1,190 @@
+local wibox    = require("wibox")
+local naughty  = require("naughty")
+local awful    = require("awful")
+local gears    = require("gears")
+
+local function factory(args)
+    local args     = args or {}
+
+    local wirelessStatus = {
+        widget    = args.widget or wibox.widget.imagebox(),
+        settings  = args.settings or function(self) end,
+        timeout   = args.timeout or 10,
+        pressed   = args.pressed or function(self, button) end,
+        followtag = args.followtag or false,
+        notification_preset = args.notification_preset or {},
+        showpopup = args.showpopup or "on",  -- Show notification popup while hovering above widget
+        timer     = gears.timer,
+        interface = "",
+        perc      = 0,
+        status    = "",
+        notification_text = "No connection",
+    }
+
+    -- Reset to default values
+    function wirelessStatus:reset()
+        wirelessStatus.interface = ""
+        wirelessStatus.perc = 0
+        wirelessStatus.status = ""
+        wirelessStatus.notification_text = "No connection"
+    end
+
+    -- Show notification popup
+    function wirelessStatus:show(seconds)
+        self:hide()
+
+        -- Update every time
+        self:update()
+
+        if self.followtag then
+            self.notification_preset.screen = focused()
+        end
+
+        self.notification = naughty.notify {
+            preset  = self.notification_preset,
+            text    = self.notification_text,
+            timeout = type(seconds) == "number" and seconds or self.notification_preset.timeout
+        }
+    end
+
+    -- Hide notification popup
+    function wirelessStatus:hide()
+        if self.notification then
+            naughty.destroy(self.notification)
+            self.notification = nil
+        end
+    end
+
+    function wirelessStatus:attach()
+        self.widget:connect_signal("mouse::enter", function()
+            self:show(0)
+        end)
+        self.widget:connect_signal("mouse::leave", function()
+            self:hide()
+        end)
+    end
+
+    -- Get name of wireless interface
+    function wirelessStatus:getInterface()
+        awful.spawn.easy_async_with_shell(
+            "awk 'NR==3 {printf(\"%s\\n\", $1)}' /proc/net/wireless",
+            function(stdout, exitcode)
+                -- Store interface name
+                -- Remove last character/s from string ("wlp4s0:"" -> "wlp4s0")
+                self.interface = stdout:sub(0, -3)
+
+                -- This is needed the first time there is an update
+                -- (or every time the `.interface` is equal to "")
+                self:update()
+            end
+        )
+    end
+
+    function wirelessStatus:update()
+        self.timer:emit_signal("timeout")
+    end
+
+    -- Get status and Quality link (convert quality link to percentages)
+    wirelessStatus, wirelessStatus.timer = awful.widget.watch(
+        "awk 'NR==3 {printf(\"%d-%.0f\\n\", $2, $3*10/7)}' /proc/net/wireless",
```

Comment:
If you want to get rid of the quoting, try this instead of the string argument:
```lua
  { 'awk', 'NR==3 {printf("%d-%.0f\n", $2, $3*10/7)}', '/proc/net/wireless' },
```
By providing the command as a list of strings like the above, the C code does not have to produce such a list by itself. Instead, the arguments are passed through to `execv()` as-is.

Reply:
I don't understand it completely, but I still changed it to what you are suggesting.

I also tried to use this for `awful.spawn.easy_async_with_shell()` but it doesn't seem to work.

Reply:
Thanks!

Let me try to explain it better:

In Unix, a command gets a list of strings as its command line arguments. Not just one single string, no, a list of strings.

When you run `foo a 'b c' "d e"` in your shell, the shell turns this string into a list of arguments like this:
- `foo`
- `a`
- `b c`
- `d e`

So you need to add quotes to things to influence how the shell turns this string into a list of strings.

And here the complication starts. When you want to start a command with the argument `foo " bar` as one argument, you need to write something like `"foo \" bar"` so that the shell produces the argument you actually want (well, just writing `'foo " bar'` would be easier, but you get the idea).

The AwesomeWM C code then has to do similar things as a shell to get all this quoting to work properly. Providing a list of strings directly is easier / more obvious, but also less obvious since it is quite uncommon to be able to do something like this.

Hopefully this explains it a bit better.

Reply:
> I also tried to use this for awful.spawn.easy_async_with_shell() but it doesn't seem to work.

Oh, I forgot to answer this part.

The `_with_shell` functions start `sh -c your-command-here`. So, a shell is started that can then do shell stuff (job management, variable substitution, whatever you need). The `your-command-here` should be a single argument, even when it contains spaces, `'` and `"`. Thus, we would have to escape the command to get the parameter through as intended.

Instead, we use the "list argument" stuff above: We start `{ 'sh', '-c', your-command-here }`. Thus, this already uses the list feature. This is why `your-command-here` must be a string and cannot be a list of strings.

Reply:
Aaaa I think I understand now. 
That list of strings is actually what is passed to scripts (python, C...) as command line arguments.

Thanks for the explanation!